### PR TITLE
Engt 1864 erc20 non transferable and redeemable

### PR DESCRIPTION
**Pull request by seansing9**
### Release Notes
[JIRA Link](https://snickerdoodlelabs.atlassian.net/browse/ENGT-1864)

#### Summary:
To support the functionality of experience points, this PR updates the ERC20Reward contract with two things: 

- Tokens minted are non-transferable by overriding the ERC20's internal function `_transfer` and adding a check that the `from` and `to` addresses by zero addresses.  Effectively, only transfers to zero addresses are allowed.

- A `redeem` function is added that takes a signature. The message contains the contract address, the address of the account redeeming tokens and the amount. After recovering the signer, if the caller of the `redeem` function has the DEFAULT_ADMIN_ROLE, we compare the signer to the target address being redeemed from. If the caller does not have the DEFAULT_ADMIN_ROLE, we compare the signer to the caller's address.

### Pre-Flight Checks
- [ ] Has QA approved this change for dev?
- [ ] Are all unit tests passing?
- [ ] Have you added this description to this week's [release notes](https://drive.google.com/drive/folders/1ELnyVZHgISIlwDQXgy0mb-4qsn_1PRZr?usp=sharing)?

### CODE REVIEW

**Review by SnickerChar — APPROVED** [review]

File: packages/contracts/contracts/testing/ERC20Reward.sol [hunk 1]
```diff
@@ -21,4 +22,46 @@ contract ERC20Reward is ERC20, ERC20Burnable, ERC20Permit, ERC7529 {
...
+
+        // Get the signer of the message
+        address signer = _getSigner(hash, signature);
+
+        // If the caller has is an admin, check the signer against the target account to burn tokens from
+        // If the caller is not an admin, check that the function caller is the signer
+        if (hasRole(DEFAULT_ADMIN_ROLE, _msgSender())) {
+            require(signer == account, "ERC20Reward: Signer does not match account to redeem from");
+        } else {
+            require(signer == _msgSender(), "ERC20Reward: Caller did not sign the message");
```
Comment (moahammadalt) [inline packages/contracts/contracts/testing/ERC20Reward.sol:41]: I am having hard times digesting this, I thought that signer == account should always be there wether the caller is admin or not and then if the caller if not an admin  _msgSender() == account is also needed as only admins can call this message to burn other accounts tokens, I may be getting this wrong though so let's discuss this off chat when you have some time @seansing9

Comment (moahammadalt) [reply packages/contracts/contracts/testing/ERC20Reward.sol:41]: The scenario running in my head is like the following:

1. Account A (the attacker) generates a signed message that authorizes a token burn from Account B (the victim). This message includes the contract address, Account B's address as the account from which tokens should be redeemed, and the amount of tokens to burn.

2. Account A then calls the redeem function with this signed message.

if the Account A is not an admin, signer == _msgSender() will get him passed as he's the account who signed the signature and called the redeem method and therefore burns tokens for Account B without being authorized.

Please let me know if I am missing something here

Comment (SnickerChar) [inline packages/contracts/contracts/testing/ERC20Reward.sol:41]: What a good catch! This needs an extra signer == _msgSender() && account == signer

**Review by SnickerChar — CHANGES_REQUESTED** [review]
Muhammed caught a nasty bug in there.

**Review by SnickerChar — DISMISSED** [review]

**Review by Tasmimdristy7 — DISMISSED** [review]

**Review by Tasmimdristy7 — APPROVED** [review]

**Pull request merged**